### PR TITLE
[MOB-16572] Retrieve last set configuration shared state to speed up Identity boot and forceSync on subsequent launch

### PR DESCRIPTION
**Pull request by addb**
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] I have signed the [Adobe Open Source CLA](http://opensource.adobe.com/cla.html).
- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Review by kevinlind — COMMENTED** [review]
Don't forget to set the resolution to `lastSet` each time you get the Configuration shared state. Currently the `lastSet` resolution is used in `readyForEvent()`, but you'll need to update each time the Configuration state is fetched when actually handling the events.

File: AEPIdentity/Tests/IdentityTests.swift [hunk 1]
```diff
@@ -485,6 +485,19 @@ class IdentityTests: XCTestCase {
         XCTAssertTrue(identity.readyForEvent(firstEvent))
     }
 
+    func testReadyForEventBootupWithValidLastSetConfigurationReturnsTrue() {
```
Comment (emdobrin) [inline AEPIdentity/Tests/IdentityTests.swift:488]: Can we also add a test to IdentityFunctionalTests.swift that checks that a sync event uses last config set?
setup:
simulate config shared state 1 orgid privacy optunknown, .set
simulate config shared state 2 orgid privacy optin, .set
simulate config shared state 3 nil, .pending
send syncIdentifier event
assert last valid config 2 used, and network call sent with orgid

File: AEPIdentity/Tests/TestHelpers/TestableExtensionRuntime.swift [hunk 2]
```diff
@@ -83,12 +83,24 @@ class TestableExtensionRuntime: ExtensionRuntime {
     }
 
     func simulateSharedState(extensionName: String, event: Event?, data: (value: [String: Any]?, status: SharedStateStatus)) {
-        otherSharedStates[extensionName] = SharedStateResult(status: data.status, value: data.value)
-        otherSharedStates["\(extensionName)-\(String(describing: event?.id))"] = SharedStateResult(status: data.status, value: data.value)
+        var sharedStateValue : [String: Any]?
```
Comment (addb) [inline AEPIdentity/Tests/TestHelpers/TestableExtensionRuntime.swift:86]: Need your review for this change @cdhoffmann.

Comment (cdhoffmann) [reply AEPIdentity/Tests/TestHelpers/TestableExtensionRuntime.swift:86]: LGTM

File: AEPIdentity/Tests/IdentityFunctionalTests.swift [hunk 3]
```diff
@@ -395,6 +422,30 @@ class IdentityFunctionalTests: XCTestCase {
         XCTAssertFalse(ecid.isEmpty)
     }
 
+    /// forseSyncIdentifier needs valid configuration to be present to process
+    func testGetECIDWithValidLastSetConfigDoesNotWaitForLatestConfigurationSharedStateToResolve() {
```
Comment (emdobrin) [inline AEPIdentity/Tests/IdentityFunctionalTests.swift:426]: simplified: testGetECIDResolvesToLastSetConfig

File: AEPIdentity/Tests/IdentityFunctionalTests.swift [hunk 4]
```diff
@@ -86,6 +86,33 @@ class IdentityFunctionalTests: XCTestCase {
         XCTAssertTrue(mockRuntime.createdSharedStates.isEmpty)
     }
 
+    static func createConfigUpdateEvent(configDict: [String: Any]) -> Event {
+        return Event(name: "Configure with Programmatic config", type: EventType.configuration, source: EventSource.requestContent,
+                     data: ["config.update": configDict])
+    }
+
+    func testSyncIdentifiersSynEvenWhenLatestConfigurationStateIsPendingtUsesLastSetConfiguration() {
```
Comment (emdobrin) [inline AEPIdentity/Tests/IdentityFunctionalTests.swift:94]: testSyncIdentifiersResolvesToLastSetConfig

File: AEPIdentity/Tests/IdentityFunctionalTests.swift [hunk 5]
```diff
@@ -86,6 +86,33 @@ class IdentityFunctionalTests: XCTestCase {
         XCTAssertTrue(mockRuntime.createdSharedStates.isEmpty)
     }
 
+    static func createConfigUpdateEvent(configDict: [String: Any]) -> Event {
+        return Event(name: "Configure with Programmatic config", type: EventType.configuration, source: EventSource.requestContent,
+                     data: ["config.update": configDict])
+    }
```
Comment (emdobrin) [inline AEPIdentity/Tests/IdentityFunctionalTests.swift:92]: let's move helper functions on top

File: AEPIdentity/Tests/IdentityTests.swift [hunk 6]
```diff
@@ -485,6 +485,18 @@ class IdentityTests: XCTestCase {
...
```
Comment (emdobrin) [inline AEPIdentity/Tests/IdentityTests.swift:488]: testReadyForEventOnBootupResolvesToLastSetConfig

**Review by emdobrin — COMMENTED** [review]
👍 Looks good with some minor renaming

File: AEPIdentity/Tests/IdentityFunctionalTests.swift (hunk 5 above)
Comment (addb) [reply AEPIdentity/Tests/IdentityFunctionalTests.swift:92]: Removing this as it is not being used.

File: AEPIdentity/Sources/Identity.swift [hunk 7]
```diff
@@ -253,7 +253,7 @@ import Foundation
     /// Sends an opt-out network request if the current privacy status is opt-out
     /// - Parameter event: the event responsible for sending this opt-out hit
     private func sendOptOutRequest(event: Event) {
-        guard let configSharedState = getSharedState(extensionName: IdentityConstants.SharedStateKeys.CONFIGURATION, event: event)?.value else { return }
+        guard let configSharedState = getSharedState(extensionName: IdentityConstants.SharedStateKeys.CONFIGURATION, event: event, resolution: .lastSet)?.value else { return }
```
[the remark this answers is not in the document]
Comment (kevinlind) [reply AEPIdentity/Sources/Identity.swift:256]: I believe this works as expected. Configuration will set its state before dispatching the configuration response event. Additionally, the shared state resolver will return the state at the given event or lower, but not higher (unless there is no state at a lower version). E.g. getting state at version 5 will not return the state at version 6.
This function is also called from `handleAudienceResponse` so the Configuration state needs to be retrieved in those cases anyway.

**Review by emdobrin — APPROVED** [review]

**Pull request merged**